### PR TITLE
Accommodate jobs larger than 256K on AWS

### PR DESCRIPTION
**Pull request by adderan**

### CODE REVIEW

**Comment by cket** [comment]
Jenkins, test this please

File: src/toil/jobStores/aws/jobStore.py [hunk 1]
```diff
@@ -342,23 +370,33 @@ def load(self, jobStoreID):
...
     def delete(self, jobStoreID):
         # remove job and replace with jobStoreId.
         log.debug("Deleting job %s", jobStoreID)
+
+        #If the job is overlarge, delete its file from the filestore
+        item = None
+        for attempt in retry_sdb():
+            with attempt:
+                item = self.jobsDomain.get_attributes(jobStoreID, consistent_read=True)
+        if self.fileExists(jobStoreID):
```
Comment (cket) [inline src/toil/jobStores/aws/jobStore.py]: Won't/Shouldn't `item` reflect whether or not the job has been written to S3?

**Comment by cket** [comment]
The test failure isn't caused by anything obvious but it has showed up twice in a row and I haven't seen it occur anywhere else, so I think it is very likely introduced in this PR.

Jenkins, test this please.

**Comment by adderan** [comment]
Thanks, agreed on all of this. Will fix.

File: src/toil/jobStores/aws/jobStore.py [hunk 2]
```diff
@@ -305,15 +305,43 @@ def _registered(self, value):
...
+                binary = fh.read()
+        else:
+            binary,_ = AWSJob.attributesToBinary(item)
+            assert binary is not None
+        job = cPickle.loads(binary)
+        return job
 
-    def create(self, jobNode):
+    def _awsJobToItem(self, job, forceOverlarge=False):
+        binary = cPickle.dumps(job)
```
Comment (joelarmstrong) [inline src/toil/jobStores/aws/jobStore.py]: This should be changed to use cPickle.HIGHEST_PROTOCOL

**Comment by cket** [comment]
This branch is still throwing errors with ``The specified domain does not exist``. Have you guys seen that in your testing?

**Comment by adderan** [comment]
Yeah I've been trying to figure this out. Since that test creates no overlarge jobs, the only code that should be executing differently is in AWSJobStore.delete, where the item is now downloaded first to check whether or not it's overlarge. 

It looks like AWSJobStore.initialize() doesn't block until the domain exists when creating the jobstore for the first time. Is there reason to believe this will always be okay?

**Comment by joelarmstrong** [comment]
FWIW I've never run into that error in any of my testing, but I think the patch that I'm using is slightly different (it's an earlier version).

**Comment by cket** [comment]
It _is_ failing at the first place where the job domain is queried after creation so that could be indicative of some sort of domain creation consistency issue. 

On the other hand the file domain is queried successfully before that call and we still haven't seen the error on any other branch, so who knows. I'll test the changes out locally tomorrow, maybe stepping through with a debugger will reveal something.

**Comment by adderan** [comment]
On the master branch I've gotten the SDB error for one of the retries:  
```
============================================= test session starts ==============================================
platform linux2 -- Python 2.7.12+, pytest-3.0.6, py-1.4.32, pluggy-0.4.0 -- /home/alden/toil/env/bin/python2
cachedir: .cache
rootdir: /home/alden/toil, inifile: setup.cfg
collected 9 items 

src/toil/test/src/fileStoreTest.py::NonCachingFileStoreTestWithAwsJobStore::testNewJobsCanHandleOtherJobDeaths thinkpad 2017-03-15 18:05:51,213 MainThread INFO toil.lib.bioio: Root logger is at level 'INFO', 'toil' logger at l
evel 'INFO'.
thinkpad 2017-03-15 18:05:51,214 MainThread INFO toil.lib.bioio: Logging to file '/tmp/toil-test-toil.test.src.fileStoreTest.NonCachingFileStoreTestWithAwsJobStore-testNewJobsCanHandleOtherJobDeaths-dGQcZn/logFile'.
thinkpad 2017-03-15 18:05:51,214 MainThread INFO toil.lib.bioio: Root logger is at level 'INFO', 'toil' logger at level 'INFO'.
thinkpad 2017-03-15 18:05:51,214 MainThread INFO toil.lib.bioio: Logging to file '/tmp/toil-test-toil.test.src.fileStoreTest.NonCachingFileStoreTestWithAwsJobStore-testNewJobsCanHandleOtherJobDeaths-dGQcZn/logFile'.
thinkpad 2017-03-15 18:05:55,368 MainThread INFO toil.jobStores.abstractJobStore: The workflow ID is: '8e0c23fb-2d79-4b3e-a31b-28e764537c95'
thinkpad 2017-03-15 18:05:55,404 MainThread INFO bd2k.util.retry: Got SDBResponseError: 400 Bad Request
<?xml version="1.0"?>
<Response><Errors><Error><Code>NoSuchDomain</Code><Message>The specified domain does not exist.</Message><BoxUsage>0.0000093202</BoxUsage></Error></Errors><RequestID>c853520d-fad7-2b29-8b1d-ae7e802d5264</RequestID></Response>,
 trying again in 0s.
thinkpad 2017-03-15 18:05:56,306 MainThread INFO toil.common: Using the single machine batch system
thinkpad 2017-03-15 18:05:56,306 MainThread WARNING toil.batchSystems.singleMachine: Limiting maxCores to CPU count of system (4).
thinkpad 2017-03-15 18:05:56,306 MainThread WARNING toil.batchSystems.singleMachine: Limiting maxMemory to physically available memory (3836411904).
thinkpad 2017-03-15 18:05:56,306 MainThread INFO toil.common: Created the workflow directory at /tmp/toil-test-toil.test.src.fileStoreTest.NonCachingFileStoreTestWithAwsJobStore-testNewJobsCanHandleOtherJobDeaths-dGQcZn/toil-8
e0c23fb-2d79-4b3e-a31b-28e764537c95
thinkpad 2017-03-15 18:05:56,307 MainThread WARNING toil.batchSystems.singleMachine: Limiting maxDisk to physically available disk (41621393408).
thinkpad 2017-03-15 18:05:56,313 MainThread INFO toil.common: User script ModuleDescriptor(dirPath='/home/alden/toil/src', name='toil.job', fromVirtualEnv=False) belongs to Toil. No need to hot-deploy it.
thinkpad 2017-03-15 18:05:56,313 MainThread INFO toil.common: No user script to hot-deploy.
thinkpad 2017-03-15 18:05:56,423 MainThread INFO toil.common: Written the environment for the jobs to the environment file
thinkpad 2017-03-15 18:05:56,423 MainThread INFO toil.common: Caching all jobs in job store
thinkpad 2017-03-15 18:05:56,747 MainThread INFO toil.common: 0 jobs downloaded.
thinkpad 2017-03-15 18:05:57,782 MainThread INFO toil: Running Toil version 3.7.0a1-dfdb3020a538a089229e4270634039272bcd3de1-dirty.
thinkpad 2017-03-15 18:05:57,782 MainThread INFO toil.realtimeLogger: Real-time logging disabled
thinkpad 2017-03-15 18:05:57,782 MainThread INFO toil.toilState: (Re)building internal scheduler state
thinkpad 2017-03-15 18:05:57,782 MainThread INFO toil.leader: Found 1 jobs to start and 0 jobs with successors to run
thinkpad 2017-03-15 18:05:57,782 MainThread INFO toil.leader: Checked batch system has no running jobs and no updated jobs
thinkpad 2017-03-15 18:05:57,783 MainThread INFO toil.leader: Starting the main loop
thinkpad 2017-03-15 18:05:57,784 MainThread INFO toil.leader: Issued job 'Job' 3d6e2e0a-6ee5-4730-a642-9850c8e5f11a with job batch system ID: 0 and cores: 1, disk: 2.0 G, and memory: 2.0 G
thinkpad 2017-03-15 18:05:58,903 MainThread INFO toil.leader: Job ended successfully: 'Job' 3d6e2e0a-6ee5-4730-a642-9850c8e5f11a
thinkpad 2017-03-15 18:05:59,068 MainThread INFO toil.leader: Issued job '_testNewJobsCanHandleOtherJobDeaths_A' 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db with job batch system ID: 1 and cores: 1, disk: 2.0 G, and memory: 100.0 M
thinkpad 2017-03-15 18:05:59,074 MainThread INFO toil.leader: Issued job '_testNewJobsCanHandleOtherJobDeaths_B' e37d343b-8981-48aa-9a15-d6dc073ca7b1 with job batch system ID: 2 and cores: 1, disk: 2.0 G, and memory: 100.0 M
Killed
thinkpad 2017-03-15 18:06:00,393 Thread-3 ERROR toil.batchSystems.singleMachine: Got exit code 137 (indicating failure) from job /home/alden/toil/env/bin/_toil_worker aws:us-west-2:cache-tests-c97faf61-b05b-4ad8-bcf3-d077520eb
5cc 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db.
thinkpad 2017-03-15 18:06:00,444 MainThread WARNING toil.leader: Job failed with exit value 137: '_testNewJobsCanHandleOtherJobDeaths_A' 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db
thinkpad 2017-03-15 18:06:00,514 MainThread WARNING toil.leader: No log file is present, despite job failing: '_testNewJobsCanHandleOtherJobDeaths_A' 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db
thinkpad 2017-03-15 18:06:00,514 MainThread WARNING toil.jobGraph: Due to failure we are reducing the remaining retry count of job '_testNewJobsCanHandleOtherJobDeaths_A' 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db with ID 49e39e5c-e
a5c-4f0d-a9fe-39d221ede9db to 0
thinkpad 2017-03-15 18:06:00,514 MainThread WARNING toil.jobGraph: We have increased the default memory of the failed job '_testNewJobsCanHandleOtherJobDeaths_A' 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db to 2147483648 bytes
thinkpad 2017-03-15 18:06:00,571 MainThread WARNING toil.leader: Job '_testNewJobsCanHandleOtherJobDeaths_A' 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db with ID 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db is completely failed
thinkpad 2017-03-15 18:06:01,086 MainThread INFO toil.leader: Job ended successfully: '_testNewJobsCanHandleOtherJobDeaths_B' e37d343b-8981-48aa-9a15-d6dc073ca7b1
thinkpad 2017-03-15 18:06:03,121 MainThread INFO toil.leader: No jobs left to run so exiting.
thinkpad 2017-03-15 18:06:03,121 MainThread INFO toil.leader: Finished the main loop
thinkpad 2017-03-15 18:06:03,121 MainThread INFO toil.serviceManager: Waiting for service manager thread to finish ...
thinkpad 2017-03-15 18:06:03,785 MainThread INFO toil.serviceManager: ... finished shutting down the service manager. Took 0.663365125656 seconds
thinkpad 2017-03-15 18:06:03,785 MainThread INFO toil.statsAndLogging: Waiting for stats and logging collator thread to finish ...
thinkpad 2017-03-15 18:06:03,998 MainThread INFO toil.statsAndLogging: ... finished collating stats and logs. Took 0.213360071182 seconds
thinkpad 2017-03-15 18:06:04,068 MainThread INFO toil.leader: Finished toil run with 2 failed jobs
thinkpad 2017-03-15 18:06:04,068 MainThread INFO toil.leader: Failed jobs at end of the run: '_testNewJobsCanHandleOtherJobDeaths_A' 49e39e5c-ea5c-4f0d-a9fe-39d221ede9db 'Job' 3d6e2e0a-6ee5-4730-a642-9850c8e5f11a
thinkpad 2017-03-15 18:06:04,142 MainThread INFO toil.common: Attempting to delete the job store
```

But I haven't gotten a full failure of the test, which I did get once on my branch. So it looks like my changes definitely make this more likely.

**Comment by cket** [comment]
Jenkins, test this please

**Comment by cket** [comment]
The plot thickets: https://github.com/BD2KGenomics/toil/blob/master/src/toil/jobStores/aws/utils.py#L232 we are actually already retrying this error for up to 5 minutes. @adderan you saw this locally right? I've been rerunning the tests on my laptop but haven't yet reproduced this error.

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
Jenkins, test this please

**Comment by cket** [comment]
@adderan @joelarmstrong it looks like the most recent commit solved the SDB issues. Are you running this code in your testing?

**Comment by adderan** [comment]
Yes, I think this fixes it. There is still a theoretical possibility of the domain not existing when we reach the jobs() call, but the retries will catch that now.

**Comment by cket** [comment]
Great, thanks for the fix!

**Pull request merged**